### PR TITLE
Update get user API

### PR DESCRIPTION
**Pull request by samueltangz**
* Return the challenge summary for the GET "/users/:id" and "/users/me" endpoints
* Use challengeLogs to keep track on the challenges solved by an user

### CODE REVIEW

File: back-end/src/api/challenges.js [hunk 1]
```diff
@@ -39,22 +39,19 @@ router.post('/:id(\\d+)/flag', authenticate, [
...
-      user.save(),
-      challenge.save()
-    ])
-    return res.status(204).end()
-  } else {
-    return res.status(400).json({
-      'error': 'Wrong flag'
-    })
-  }
+  const challengeLog = await db.ChallengeLogs.findOne({ userId: user.id, challengeId: id })
```
Comment (harrier-lcc) [inline back-end/src/api/challenges.js]: this db call is not necessary...?

File: back-end/src/helpers/marshaler.js [hunk 2]
```diff
@@ -25,10 +25,10 @@ const marshalAnnouncement = verboseLevel => announcement => {
 }
 
 const marshalChallengeLog = verboseLevel => challengeLog => {
-  // TODO: pick something to marshal...
   switch (verboseLevel) {
     case verboseLevels.MINIMAL:
-      return _.pick(challengeLog, [ 'id' ])
+      challengeLog.challenge = marshalChallenge(verboseLevels.MINIMAL)(challengeLog.challenge)
+      return _.pick(challengeLog, [ 'challenge', 'date' ])
```
Comment (harrier-lcc) [inline back-end/src/helpers/marshaler.js]: Why only change verboseLevels.MINIMAL then?

File: back-end/src/api/challenges.js (hunk 1 above)
Comment (samueltangz) [reply back-end/src/api/challenges.js]: It is. The `challengesSolved` attribute in `users` is removed now. We use the `ChallengeLogs` model to keeep track on what challenges a user has done.

File: back-end/src/models/users.js [hunk 3]
```diff
@@ -42,25 +42,17 @@ const User = new Schema({
...
 })
 
-// Lists the users with rank between [skip, skip+limit) (zero indexed)
-User.statics.listWithRanks = async function (skip, limit) {
+// Find the users with rank (sorted by score) between [skip, skip+limit) (zero indexed)
+User.statics.findWithRanking = async function (skip, limit) {
   const users = await this.find()
-    .sort({ score: -1, lastSolveAt: 1, name: 1 })
-    .skip(skip).limit(limit)
+    .sort({ score: -1, lastSolveAt: 1 })
```
Comment (harrier-lcc) [inline back-end/src/models/users.js]: Why need to change this? The original sorting logic is, once lastSolveAt is exactly the same unfortunately, just sort by name

File: back-end/src/helpers/marshaler.js (hunk 2 above)
Comment (harrier-lcc) [inline back-end/src/helpers/marshaler.js]: How about other marshal level?

Comment (samueltangz) [reply back-end/src/helpers/marshaler.js]: They are not being used yet. Write when they are needed.

**Pull request merged**